File: polymarket/backtesting/performance_metrics.py
"""
Performance metrics and analysis for backtesting results
"""

import logging
import numpy as np
import pandas as pd
from typing import Dict, List, Tuple

logger = logging.getLogger(__name__)


class PerformanceAnalyzer:
    """Computes comprehensive performance metrics for backtest results"""

    @staticmethod
    def compute_metrics(trades: pd.DataFrame, portfolio_values: List[float],
                        initial_capital: float) -> Dict:
        """
        Compute full suite of performance metrics

        Args:
            trades: DataFrame with trade records
            portfolio_values: List of daily portfolio values
            initial_capital: Starting capital

        Returns:
            Dictionary with all performance metrics
        """
        metrics = {}

        # Portfolio-level metrics
        metrics["portfolio"] = PerformanceAnalyzer._portfolio_metrics(
            portfolio_values, initial_capital
        )

        # Trade-level metrics
        if not trades.empty:
            metrics["trades"] = PerformanceAnalyzer._trade_metrics(trades)
        else:
            metrics["trades"] = {"total_trades": 0}

        # Forecast accuracy (if outcome data available)
        if "our_estimate" in trades.columns and "actual_outcome" in trades.columns:
            metrics["forecast"] = PerformanceAnalyzer._forecast_metrics(trades)

        return metrics

    @staticmethod
    def _portfolio_metrics(portfolio_values: List[float],
                           initial_capital: float) -> Dict:
        """Compute portfolio-level performance metrics"""
        values = np.array(portfolio_values)
        if len(values) < 2:
            return {
                "initial_capital": initial_capital,
                "final_value": values[-1] if len(values) > 0 else initial_capital,
                "total_return_pct": 0,
            }

        final_value = values[-1]
        total_return = (final_value - initial_capital) / initial_capital

        # Daily returns
        daily_returns = np.diff(values) / values[:-1]
        daily_returns = daily_returns[np.isfinite(daily_returns)]

        # Sharpe ratio (annualized, assuming 365 trading days for crypto/prediction markets)
        sharpe = 0
        if len(daily_returns) > 1 and np.std(daily_returns) > 0:
            sharpe = np.mean(daily_returns) / np.std(daily_returns) * np.sqrt(365)

        # Max drawdown
        running_max = np.maximum.accumulate(values)
        drawdowns = (values - running_max) / running_max
        max_drawdown = float(np.min(drawdowns))

        # Sortino ratio
        downside_returns = daily_returns[daily_returns < 0]
        sortino = 0
        if len(downside_returns) > 0 and np.std(downside_returns) > 0:
            sortino = np.mean(daily_returns) / np.std(downside_returns) * np.sqrt(365)

        return {
            "initial_capital": initial_capital,
            "final_value": float(final_value),
            "total_return_pct": float(total_return * 100),
            "total_pnl": float(final_value - initial_capital),
            "sharpe_ratio": float(sharpe),
            "sortino_ratio": float(sortino),
            "max_drawdown_pct": float(max_drawdown * 100),
            "volatility_annual_pct": float(np.std(daily_returns) * np.sqrt(365) * 100) if len(daily_returns) > 0 else 0,
            "best_day_pct": float(np.max(daily_returns) * 100) if len(daily_returns) > 0 else 0,
            "worst_day_pct": float(np.min(daily_returns) * 100) if len(daily_returns) > 0 else 0,
            "positive_days": int(np.sum(daily_returns > 0)),
            "negative_days": int(np.sum(daily_returns < 0)),
            "total_days": len(daily_returns),
        }

    @staticmethod
    def _trade_metrics(trades: pd.DataFrame) -> Dict:
        """Compute trade-level performance metrics"""
        total = len(trades)
        if total == 0:
            return {"total_trades": 0}

        pnls = trades["pnl"].values if "pnl" in trades.columns else np.array([])

        wins = pnls[pnls > 0] if len(pnls) > 0 else np.array([])
        losses = pnls[pnls < 0] if len(pnls) > 0 else np.array([])

        win_rate = len(wins) / total if total > 0 else 0
        avg_win = float(np.mean(wins)) if len(wins) > 0 else 0
        avg_loss = float(np.mean(losses)) if len(losses) > 0 else 0
        profit_factor = abs(np.sum(wins) / np.sum(losses)) if len(losses) > 0 and np.sum(losses) != 0 else float("inf")

        return {
            "total_trades": total,
            "winning_trades": len(wins),
            "losing_trades": len(losses),
            "win_rate_pct": float(win_rate * 100),
            "avg_win": avg_win,
            "avg_loss": avg_loss,
            "total_pnl": float(np.sum(pnls)) if len(pnls) > 0 else 0,
            "profit_factor": float(profit_factor),
            "largest_win": float(np.max(wins)) if len(wins) > 0 else 0,
            "largest_loss": float(np.min(losses)) if len(losses) > 0 else 0,
            "avg_trade_pnl": float(np.mean(pnls)) if len(pnls) > 0 else 0,
        }

    @staticmethod
    def _forecast_metrics(trades: pd.DataFrame) -> Dict:
        """Compute forecast accuracy metrics"""
        predictions = trades["our_estimate"].values
        outcomes = trades["actual_outcome"].values

        # Brier score
        brier_score = float(np.mean((predictions - outcomes) ** 2))

        # Log loss
        eps = 1e-15
        clipped = np.clip(predictions, eps, 1 - eps)
        log_loss = -float(np.mean(
            outcomes * np.log(clipped) + (1 - outcomes) * np.log(1 - clipped)
        ))

        # Calibration (binned)
        calibration = PerformanceAnalyzer._calibration_curve(predictions, outcomes)

        # Expected Calibration Error
        ece = PerformanceAnalyzer._expected_calibration_error(predictions, outcomes)

        return {
            "brier_score": brier_score,
            "log_loss": log_loss,
            "expected_calibration_error": ece,
            "calibration_bins": calibration,
        }

    @staticmethod
    def _calibration_curve(
        predictions: np.ndarray, outcomes: np.ndarray, n_bins: int = 10
    ) -> List[Dict]:
        """Compute calibration curve data"""
        bins = np.linspace(0, 1, n_bins + 1)
        curve = []

        for i in range(n_bins):
            mask = (predictions >= bins[i]) & (predictions < bins[i + 1])
            if np.sum(mask) > 0:
                curve.append({
                    "bin_center": float((bins[i] + bins[i + 1]) / 2),
                    "predicted_mean": float(np.mean(predictions[mask])),
                    "observed_freq": float(np.mean(outcomes[mask])),
                    "count": int(np.sum(mask)),
                })

        return curve

    @staticmethod
    def _expected_calibration_error(
        predictions: np.ndarray, outcomes: np.ndarray, n_bins: int = 10
    ) -> float:
        """Compute Expected Calibration Error"""
        bins = np.linspace(0, 1, n_bins + 1)
        ece = 0.0
        total = len(predictions)

        for i in range(n_bins):
            mask = (predictions >= bins[i]) & (predictions < bins[i + 1])
            count = np.sum(mask)
            if count > 0:
                avg_pred = np.mean(predictions[mask])
                avg_outcome = np.mean(outcomes[mask])
                ece += (count / total) * abs(avg_pred - avg_outcome)

        return float(ece)

    @staticmethod
    def compute_city_breakdown(
        all_trades: pd.DataFrame,
    ) -> Dict[str, Dict]:
        """Compute per-city performance breakdown"""
        if "city" not in all_trades.columns:
            return {}

        breakdown = {}
        for city in all_trades["city"].unique():
            city_trades = all_trades[all_trades["city"] == city]
            pnls = city_trades["pnl"].values if "pnl" in city_trades.columns else np.array([])
            wins = pnls[pnls > 0] if len(pnls) > 0 else np.array([])

            breakdown[city] = {
                "total_trades": len(city_trades),
                "total_pnl": float(np.sum(pnls)) if len(pnls) > 0 else 0,
                "win_rate_pct": float(len(wins) / len(city_trades) * 100) if len(city_trades) > 0 else 0,
                "avg_trade_pnl": float(np.mean(pnls)) if len(pnls) > 0 else 0,
            }

        return breakdown

    @staticmethod
    def compute_market_type_breakdown(
        all_trades: pd.DataFrame,
    ) -> Dict[str, Dict]:
        """Compute per-market-type performance breakdown"""
        if "market_type" not in all_trades.columns:
            return {}

        breakdown = {}
        for mtype in all_trades["market_type"].unique():
            type_trades = all_trades[all_trades["market_type"] == mtype]
            pnls = type_trades["pnl"].values if "pnl" in type_trades.columns else np.array([])
            wins = pnls[pnls > 0] if len(pnls) > 0 else np.array([])

            breakdown[mtype] = {
                "total_trades": len(type_trades),
                "total_pnl": float(np.sum(pnls)) if len(pnls) > 0 else 0,
                "win_rate_pct": float(len(wins) / len(type_trades) * 100) if len(type_trades) > 0 else 0,
                "avg_trade_pnl": float(np.mean(pnls)) if len(pnls) > 0 else 0,
            }

        return breakdown

    @staticmethod
    def compute_monthly_breakdown(
        all_trades: pd.DataFrame,
    ) -> Dict[str, Dict]:
        """Compute monthly performance breakdown"""
        if "date" not in all_trades.columns or all_trades.empty:
            return {}

        all_trades = all_trades.copy()
        all_trades["month"] = pd.to_datetime(all_trades["date"]).dt.to_period("M").astype(str)

        breakdown = {}
        for month in sorted(all_trades["month"].unique()):
            month_trades = all_trades[all_trades["month"] == month]
            pnls = month_trades["pnl"].values if "pnl" in month_trades.columns else np.array([])
            wins = pnls[pnls > 0] if len(pnls) > 0 else np.array([])

            breakdown[month] = {
                "total_trades": len(month_trades),
                "total_pnl": float(np.sum(pnls)) if len(pnls) > 0 else 0,
                "win_rate_pct": float(len(wins) / len(month_trades) * 100) if len(month_trades) > 0 else 0,
            }

        return breakdown

File: src/backtesting/performance_metrics.rs
use std::collections::HashMap;

use chrono::NaiveDate;
use serde::{Deserialize, Serialize};

use crate::types::TradeRecord;
use crate::utils::{max_drawdown, mean, std_dev};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PortfolioMetrics {
    pub initial_capital: f64,
    pub final_value: f64,
    pub total_return_pct: f64,
    pub total_pnl: f64,
    pub sharpe_ratio: f64,
    pub sortino_ratio: f64,
    pub max_drawdown_pct: f64,
    pub volatility_annual_pct: f64,
    pub best_day_pct: f64,
    pub worst_day_pct: f64,
    pub positive_days: usize,
    pub negative_days: usize,
    pub total_days: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TradeMetrics {
    pub total_trades: usize,
    pub winning_trades: usize,
    pub losing_trades: usize,
    pub win_rate_pct: f64,
    pub avg_win: f64,
    pub avg_loss: f64,
    pub total_pnl: f64,
    pub profit_factor: f64,
    pub largest_win: f64,
    pub largest_loss: f64,
    pub avg_trade_pnl: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ForecastMetrics {
    pub brier_score: f64,
    pub log_loss: f64,
    pub expected_calibration_error: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PerformanceMetrics {
    pub portfolio: PortfolioMetrics,
    pub trades: TradeMetrics,
    pub forecast: Option<ForecastMetrics>,
    pub by_city: HashMap<String, HashMap<String, f64>>,
    pub by_market_type: HashMap<String, HashMap<String, f64>>,
    pub by_month: HashMap<String, HashMap<String, f64>>,
}

pub struct PerformanceAnalyzer;

impl PerformanceAnalyzer {
    pub fn compute_metrics(
        trades: &[TradeRecord],
        portfolio_values: &[f64],
        initial_capital: f64,
    ) -> PerformanceMetrics {
        let portfolio = Self::portfolio_metrics(portfolio_values, initial_capital);
        let trade_metrics = Self::trade_metrics(trades);

        let forecast = if trades.is_empty() {
            None
        } else {
            Some(Self::forecast_metrics(trades))
        };

        PerformanceMetrics {
            portfolio,
            trades: trade_metrics,
            forecast,
            by_city: Self::compute_city_breakdown(trades),
            by_market_type: Self::compute_market_type_breakdown(trades),
            by_month: Self::compute_monthly_breakdown(trades),
        }
    }

    pub fn portfolio_metrics(portfolio_values: &[f64], initial_capital: f64) -> PortfolioMetrics {
        if portfolio_values.len() < 2 {
            return PortfolioMetrics {
                initial_capital,
                final_value: *portfolio_values.last().unwrap_or(&initial_capital),
                total_return_pct: 0.0,
                total_pnl: 0.0,
                sharpe_ratio: 0.0,
                sortino_ratio: 0.0,
                max_drawdown_pct: 0.0,
                volatility_annual_pct: 0.0,
                best_day_pct: 0.0,
                worst_day_pct: 0.0,
                positive_days: 0,
                negative_days: 0,
                total_days: 0,
            };
        }

        let final_value = *portfolio_values.last().unwrap_or(&initial_capital);
        let total_return = (final_value - initial_capital) / initial_capital;
        let daily_returns: Vec<f64> = portfolio_values
            .windows(2)
            .filter_map(|w| if w[0] != 0.0 { Some((w[1] - w[0]) / w[0]) } else { None })
            .collect();

        let sharpe = {
            let sd = std_dev(&daily_returns, 0);
            if daily_returns.len() > 1 && sd > 0.0 {
                mean(&daily_returns) / sd * 365.0_f64.sqrt()
            } else {
                0.0
            }
        };

        let downside: Vec<f64> = daily_returns.iter().copied().filter(|r| *r < 0.0).collect();
        let sortino = {
            let sd = std_dev(&downside, 0);
            if !downside.is_empty() && sd > 0.0 {
                mean(&daily_returns) / sd * 365.0_f64.sqrt()
            } else {
                0.0
            }
        };

        PortfolioMetrics {
            initial_capital,
            final_value,
            total_return_pct: total_return * 100.0,
            total_pnl: final_value - initial_capital,
            sharpe_ratio: sharpe,
            sortino_ratio: sortino,
            max_drawdown_pct: max_drawdown(portfolio_values) * 100.0,
            volatility_annual_pct: std_dev(&daily_returns, 0) * 365.0_f64.sqrt() * 100.0,
            best_day_pct: daily_returns
                .iter()
                .copied()
                .reduce(f64::max)
                .unwrap_or(0.0)
                * 100.0,
            worst_day_pct: daily_returns
                .iter()
                .copied()
                .reduce(f64::min)
                .unwrap_or(0.0)
                * 100.0,
            positive_days: daily_returns.iter().filter(|r| **r > 0.0).count(),
            negative_days: daily_returns.iter().filter(|r| **r < 0.0).count(),
            total_days: daily_returns.len(),
        }
    }

    pub fn trade_metrics(trades: &[TradeRecord]) -> TradeMetrics {
        if trades.is_empty() {
            return TradeMetrics {
                total_trades: 0,
                winning_trades: 0,
                losing_trades: 0,
                win_rate_pct: 0.0,
                avg_win: 0.0,
                avg_loss: 0.0,
                total_pnl: 0.0,
                profit_factor: f64::INFINITY,
                largest_win: 0.0,
                largest_loss: 0.0,
                avg_trade_pnl: 0.0,
            };
        }

        let pnls: Vec<f64> = trades.iter().map(|t| t.pnl).collect();
        let wins: Vec<f64> = pnls.iter().copied().filter(|p| *p > 0.0).collect();
        let losses: Vec<f64> = pnls.iter().copied().filter(|p| *p < 0.0).collect();
        let win_rate = wins.len() as f64 / trades.len() as f64;

        let profit_factor = if losses.is_empty() || losses.iter().sum::<f64>() == 0.0 {
            f64::INFINITY
        } else {
            (wins.iter().sum::<f64>() / losses.iter().sum::<f64>()).abs()
        };

        TradeMetrics {
            total_trades: trades.len(),
            winning_trades: wins.len(),
            losing_trades: losses.len(),
            win_rate_pct: win_rate * 100.0,
            avg_win: mean(&wins),
            avg_loss: mean(&losses),
            total_pnl: pnls.iter().sum(),
            profit_factor,
            largest_win: wins.iter().copied().reduce(f64::max).unwrap_or(0.0),
            largest_loss: losses.iter().copied().reduce(f64::min).unwrap_or(0.0),
            avg_trade_pnl: mean(&pnls),
        }
    }

    pub fn forecast_metrics(trades: &[TradeRecord]) -> ForecastMetrics {
        let preds: Vec<f64> = trades.iter().map(|t| t.our_estimate).collect();
        let outs: Vec<f64> = trades.iter().map(|t| t.actual_outcome).collect();

        let brier = preds
            .iter()
            .zip(outs.iter())
            .map(|(p, y)| {
                let d = p - y;
                d * d
            })
            .sum::<f64>()
            / preds.len().max(1) as f64;

        let eps = 1e-15;
        let log_loss = -preds
            .iter()
            .zip(outs.iter())
            .map(|(p, y)| {
                let clipped = p.max(eps).min(1.0 - eps);
                y * clipped.ln() + (1.0 - y) * (1.0 - clipped).ln()
            })
            .sum::<f64>()
            / preds.len().max(1) as f64;

        let ece = expected_calibration_error(&preds, &outs, 10);

        ForecastMetrics {
            brier_score: brier,
            log_loss,
            expected_calibration_error: ece,
        }
    }

    pub fn compute_city_breakdown(trades: &[TradeRecord]) -> HashMap<String, HashMap<String, f64>> {
        let mut buckets: HashMap<String, Vec<&TradeRecord>> = HashMap::new();
        for trade in trades {
            buckets.entry(trade.city.clone()).or_default().push(trade);
        }
        summarize_buckets(buckets)
    }

    pub fn compute_market_type_breakdown(
        trades: &[TradeRecord],
    ) -> HashMap<String, HashMap<String, f64>> {
        let mut buckets: HashMap<String, Vec<&TradeRecord>> = HashMap::new();
        for trade in trades {
            buckets
                .entry(trade.market_type.clone())
                .or_default()
                .push(trade);
        }
        summarize_buckets(buckets)
    }

    pub fn compute_monthly_breakdown(trades: &[TradeRecord]) -> HashMap<String, HashMap<String, f64>> {
        let mut buckets: HashMap<String, Vec<&TradeRecord>> = HashMap::new();
        for trade in trades {
            let month = trade.date.format("%Y-%m").to_string();
            buckets.entry(month).or_default().push(trade);
        }
        summarize_buckets(buckets)
    }
}

fn summarize_buckets(
    buckets: HashMap<String, Vec<&TradeRecord>>,
) -> HashMap<String, HashMap<String, f64>> {
    let mut out = HashMap::new();

    for (k, trades) in buckets {
        let pnls: Vec<f64> = trades.iter().map(|t| t.pnl).collect();
        let wins = pnls.iter().filter(|p| **p > 0.0).count() as f64;
        let total = trades.len().max(1) as f64;

        let mut row = HashMap::new();
        row.insert("total_trades".to_string(), trades.len() as f64);
        row.insert("total_pnl".to_string(), pnls.iter().sum());
        row.insert("win_rate_pct".to_string(), wins / total * 100.0);
        row.insert("avg_trade_pnl".to_string(), mean(&pnls));
        out.insert(k, row);
    }

    out
}

fn expected_calibration_error(predictions: &[f64], outcomes: &[f64], n_bins: usize) -> f64 {
    if predictions.is_empty() || predictions.len() != outcomes.len() {
        return 0.0;
    }

    let mut ece = 0.0;
    for i in 0..n_bins {
        let lo = i as f64 / n_bins as f64;
        let hi = (i + 1) as f64 / n_bins as f64;

        let mut p_bin = Vec::new();
        let mut y_bin = Vec::new();
        for (p, y) in predictions.iter().zip(outcomes.iter()) {
            if *p >= lo && *p < hi {
                p_bin.push(*p);
                y_bin.push(*y);
            }
        }

        if !p_bin.is_empty() {
            let w = p_bin.len() as f64 / predictions.len() as f64;
            ece += w * (mean(&p_bin) - mean(&y_bin)).abs();
        }
    }

    ece
}

pub fn make_trade_record(
    date: NaiveDate,
    city: &str,
    market_type: &str,
    pnl: f64,
    our_estimate: f64,
    actual_outcome: f64,
) -> TradeRecord {
    TradeRecord {
        date,
        city: city.to_string(),
        market_id: format!("{city}_{market_type}_{}", date.format("%Y%m%d")),
        market_title: format!("Synthetic {market_type} market"),
        market_type: market_type.to_string(),
        threshold: 0.0,
        market_price: 0.5,
        our_estimate,
        edge: our_estimate - 0.5,
        side: if our_estimate >= 0.5 {
            "BUY".to_string()
        } else {
            "SELL".to_string()
        },
        position_size: 100.0,
        actual_outcome,
        pnl,
        capital_after: 100_000.0 + pnl,
    }
}
